File: tests/occupancy-stress.test.ts
/**
 * space_occupancy concurrency stress test — TypeScript port of prototype
 * Proves the claim-range constraint under real concurrent load.
 *
 * Tests:
 *   T1  Exclusive integrity: 50 clients race for 1 room → exactly 1 winner
 *   T2  Composite exclusivity: private vs beds on same space → never both
 *   T3  Capacity integrity: 40 clients race for 6 beds → exactly 6 winners
 *   T4  Choke point: direct INSERT as app_role → 42501 permission denied
 *   T5  Throughput: 500 non-overlapping bookings, 10 threads → >1000 commits/sec
 *
 * Run: bun test src/stress/occupancy-stress.test.ts
 * Requires: local PG16 with schema loaded, seed fixture applied
 */

import { describe, expect, test, beforeAll, afterAll } from "bun:test";
import { sql } from "../db"; // your drizzle/raw SQL connection

const TENANT = "00000000-0000-0000-0000-000000000001";
const ROOM_101 = "00000000-0000-0000-0000-000000000200"; // STD, capacity 1, exclusive
const DORM_A = "00000000-0000-0000-0000-000000000215";   // 6-bed dorm (seed separately if needed)

// Helper: attempt occupancy via choke point
async function attemptOccupancy(
  spaceId: string,
  period: string,
  exclusive: boolean,
  slotKind: string = "segment"
): Promise<{ success: boolean; error?: string; occupancyId?: string }> {
  const slotRef = crypto.randomUUID();
  try {
    const result = await sql`SELECT record_occupancy(
      ${TENANT}::uuid,
      ${spaceId}::uuid,
      ${period}::tstzrange,
      ${slotRef}::uuid,
      ${slotKind},
      ${exclusive}
    ) as id`;
    return { success: true, occupancyId: result[0].id };
  } catch (e: any) {
    return { success: false, error: e.code || e.message };
  }
}

// Helper: release occupancy
async function releaseOccupancy(slotRef: string): Promise<number> {
  const result = await sql`SELECT release_occupancy(${TENANT}::uuid, ${slotRef}::uuid) as n`;
  return result[0].n;
}

// Helper: race N clients for the same space/period
async function race(
  n: number,
  spaceId: string,
  period: string,
  exclusive: boolean
): Promise<{ wins: number; rejects: number; durationMs: number; results: string[] }> {
  const results: string[] = new Array(n).fill("pending");
  const start = performance.now();

  const promises = Array.from({ length: n }, (_, i) =>
    attemptOccupancy(spaceId, period, exclusive).then((r) => {
      results[i] = r.success ? "WIN" : `REJECT:${r.error}`;
    })
  );

  await Promise.all(promises);
  const durationMs = performance.now() - start;

  const wins = results.filter((r) => r === "WIN").length;
  const rejects = n - wins;

  return { wins, rejects, durationMs, results };
}

describe("Occupancy Concurrency Stress Tests", () => {
  beforeAll(async () => {
    // Ensure clean state
    await sql`SELECT set_config('app.tenant_id', ${TENANT}, true)`;
    // Clean up any test occupancy from previous runs
    await sql`DELETE FROM space_occupancy WHERE tenant_id = ${TENANT}::uuid AND slot_kind = 'segment' AND slot_ref LIKE 'test-%'`;
  });

  afterAll(async () => {
    // Cleanup
    await sql`DELETE FROM space_occupancy WHERE tenant_id = ${TENANT}::uuid AND slot_kind = 'segment' AND slot_ref LIKE 'test-%'`;
  });

  test("T1: Exclusive race — 50 clients, 1 room, exactly 1 winner", async () => {
    const period = "[2026-10-01 14:00+04,2026-10-03 12:00+04)";
    const { wins, rejects, durationMs } = await race(50, ROOM_101, period, true);

    console.log(`T1: wins=${wins}, rejects=${rejects}, ${durationMs.toFixed(0)}ms`);
    expect(wins).toBe(1);
    expect(rejects).toBe(49);
    expect(durationMs).toBeLessThan(5000); // Should complete in <5s
  });

  test("T2: Composite race — private vs beds, never both", async () => {
    // This test requires a dorm space. If not seeded, skip.
    const dormCheck = await sql`SELECT id FROM space WHERE id = ${DORM_A}::uuid`;
    if (dormCheck.length === 0) {
      console.log("T2: SKIP — dorm space not seeded");
      return;
    }

    const period = "[2026-10-05 14:00+04,2026-10-07 12:00+04)";
    const results: string[] = new Array(50).fill("pending");

    const promises = Array.from({ length: 50 }, (_, i) => {
      const exclusive = i % 2 === 0; // 25 exclusive, 25 positional
      return attemptOccupancy(DORM_A, period, exclusive).then((r) => {
        results[i] = r.success ? (exclusive ? "EXCL_WIN" : "BED_WIN") : `REJECT:${r.error}`;
      });
    });

    await Promise.all(promises);

    const exclWins = results.filter((r) => r === "EXCL_WIN").length;
    const bedWins = results.filter((r) => r === "BED_WIN").length;

    console.log(`T2: private=${exclWins}, beds=${bedWins}`);

    // Legal outcomes: (1,0) or (0,1..6) — NEVER both
    const validOutcome =
      (exclWins === 1 && bedWins === 0) ||
      (exclWins === 0 && bedWins >= 1 && bedWins <= 6);

    expect(validOutcome).toBe(true);
  });

  test("T3: Capacity race — 40 clients, 6 beds, exactly 6 winners", async () => {
    const dormCheck = await sql`SELECT id FROM space WHERE id = ${DORM_A}::uuid`;
    if (dormCheck.length === 0) {
      console.log("T3: SKIP — dorm space not seeded");
      return;
    }

    const period = "[2026-10-10 14:00+04,2026-10-12 12:00+04)";
    const { wins, rejects, durationMs } = await race(40, DORM_A, period, false);

    console.log(`T3: wins=${wins}, rejects=${rejects}, ${durationMs.toFixed(0)}ms`);
    expect(wins).toBe(6);
    expect(rejects).toBe(34);
  });

  test("T4: Choke point — direct INSERT as app_role must fail with 42501", async () => {
    try {
      await sql`INSERT INTO space_occupancy (
        tenant_id, space_id, period, slot_ref, slot_kind, exclusive, claim
      ) VALUES (
        ${TENANT}::uuid,
        ${ROOM_101}::uuid,
        '[2027-01-01,2027-01-02)'::tstzrange,
        ${crypto.randomUUID()}::uuid,
        'segment',
        true,
        int4range(0, NULL)
      )`;

      // If we get here, the INSERT succeeded — this is a FAILURE
      expect.fail("Direct INSERT should have been denied");
    } catch (e: any) {
      console.log(`T4: denied with code ${e.code}`);
      expect(e.code).toBe("42501"); // insufficient_privilege
    }
  });

  test("T5: Throughput — 500 non-overlapping bookings, 10 concurrent workers", async () => {
    const workerCount = 10;
    const bookingsPerWorker = 50;
    const results: number[] = new Array(workerCount).fill(0);

    const start = performance.now();

    const workers = Array.from({ length: workerCount }, async (_, k) => {
      let count = 0;
      for (let j = 0; j < bookingsPerWorker; j++) {
        const d = k * bookingsPerWorker + j;
        const hour = Math.floor(d / 60);
        const minute = d % 60;
        const period = `[2028-01-01 ${hour.toString().padStart(2, "0")}:${minute.toString().padStart(2, "0")}+04,2028-01-01 ${hour.toString().padStart(2, "0")}:${minute.toString().padStart(2, "0")}:30+04)`;

        const r = await attemptOccupancy(ROOM_101, period, true);
        if (r.success) count++;
      }
      results[k] = count;
    });

    await Promise.all(workers);
    const duration = (performance.now() - start) / 1000;
    const total = results.reduce((a, b) => a + b, 0);
    const rate = total / duration;

    console.log(`T5: ${total} bookings / ${duration.toFixed(2)}s = ${rate.toFixed(0)} commits/sec`);

    expect(total).toBe(workerCount * bookingsPerWorker); // All should succeed (non-overlapping)
    expect(rate).toBeGreaterThan(500); // Should sustain >500 commits/sec
  });

  test("RLS isolation — cross-tenant read returns zero rows", async () => {
    // Set tenant to a non-existent UUID
    await sql`SELECT set_config('app.tenant_id', '99999999-9999-9999-9999-999999999999', true)`;

    const spaces = await sql`SELECT * FROM space WHERE property_node = ${"00000000-0000-0000-0000-000000000012"}::uuid`;
    expect(spaces.length).toBe(0);

    // Reset
    await sql`SELECT set_config('app.tenant_id', ${TENANT}, true)`;
  });

  test("Journal balance — unbalanced journal rejected at commit", async () => {
    // Start a transaction
    const tx = await sql.begin(async (sql) => {
      // Set tenant within transaction
      await sql`SELECT set_config('app.tenant_id', ${TENANT}, true)`;

      // Create a journal
      const [journal] = await sql`
        INSERT INTO journal (tenant_id, property_node, business_date, kind, description, currency, source)
        VALUES (${TENANT}::uuid, ${"00000000-0000-0000-0000-000000000012"}::uuid, '2026-09-15', 'charge', 'Test unbalanced', 'AED', '{"test":true}')
        RETURNING id
      `;

      // Post only ONE line (unbalanced — no credit side)
      await sql`
        INSERT INTO posting_line (tenant_id, journal_id, seq, account_id, tx_code, amount_minor, business_date)
        VALUES (${TENANT}::uuid, ${journal.id}::uuid, 1, ${"00000000-0000-0000-0000-000000000800"}::uuid, 'ROOM', 50000, '2026-09-15')
      `;

      // The deferred trigger should fire at commit and reject
      // If we reach here, the trigger didn't fire — but it should on COMMIT
    });

    // If we get here without error, the test failed
    expect.fail("Unbalanced journal should have been rejected");
  });
});

/**
 * Additional integration tests for the complete guest journey
 */
describe("Guest Journey Integration Tests", () => {
  test("End-to-end: Search → Hold → Commit → Check-in → Post → Checkout", async () => {
    // This is a skeleton — implement with your actual service layer
    // 1. Search availability
    // 2. Create hold
    // 3. Commit reservation
    // 4. Verify reservation + folio + outbox events
    // 5. Check-in
    // 6. Post room charge
    // 7. Verify journal balance = 0
    // 8. Checkout
    // 9. Verify folio settled + closed
    // 10. Verify occupancy released
  });

  test("Cancellation with penalty: late cancel charges first night", async () => {
    // 1. Create reservation for tomorrow
    // 2. Cancel today (within 24h)
    // 3. Verify penalty journal posted
    // 4. Verify journal balance = 0
    // 5. Verify folio shows charge
  });

  test("No-show processing: auto-charge and release", async () => {
    // 1. Create reservation for yesterday
    // 2. Run no-show job
    // 3. Verify status = no_show
    // 4. Verify occupancy released
    // 5. Verify no-show charge posted
  });

  test("Document issuance: gapless numbering + hash chain", async () => {
    // 1. Settle folio
    // 2. Issue invoice
    // 3. Verify doc_no = INV-DXB-1
    // 4. Verify sha256 computed
    // 5. Issue second invoice
    // 6. Verify doc_no = INV-DXB-2, prev_hash = first invoice sha256
  });

  test("Business day seal: blocks new postings", async () => {
    // 1. Seal 2026-09-15
    // 2. Attempt normal charge → should fail with P0011
    // 3. Attempt adjustment → should succeed
  });
});

File: prototype/occupancy_stress_test.py
#!/usr/bin/env python3
"""
space_occupancy concurrency stress test — the one architectural claim that needed
evidence rather than reasoning.

Proves, under real concurrent load:
  T1  Exclusive integrity: N clients race for the same private room, same dates
      -> exactly 1 wins, N-1 get a clean constraint rejection.
  T2  Composite exclusivity: private-room sale vs dorm-bed sales on the SAME space
      -> whoever commits first blocks the other configuration.
  T3  Capacity integrity: N clients race for beds in a 6-bed dorm, same dates
      -> exactly 6 win.
  T4  Choke point: the app role attempting direct INSERT on space_occupancy
      -> permission denied. Only record_occupancy() works.
  T5  Throughput: sustained non-overlapping bookings/sec through the choke point.

Run:  python3 occupancy_stress_test.py  (expects local PG16 with prototype schema)
"""
import psycopg2, threading, time, uuid, random

DSN = "host=127.0.0.1 dbname=pms_proto user=app_role password=app"
TEN = '00000000-0000-0000-0000-000000000001'
ROOM = '00000000-0000-0000-0000-00000000000a'   # capacity 1, sold exclusive
DORM = '00000000-0000-0000-0000-00000000000b'   # capacity 6, beds non-exclusive

def attempt(space, period, exclusive, results, idx):
    try:
        con = psycopg2.connect(DSN); con.autocommit = False
        cur = con.cursor()
        cur.execute("SELECT record_occupancy(%s,%s,%s::tstzrange,%s,%s)",
                    (TEN, space, period, str(uuid.uuid4()), exclusive))
        con.commit(); results[idx] = 'WIN'
    except psycopg2.Error as e:
        results[idx] = ('REJECT:' + (e.pgcode or '?'))
    finally:
        try: con.close()
        except: pass

def race(n, space, period, exclusive):
    results = [None]*n
    ts = [threading.Thread(target=attempt, args=(space, period, exclusive, results, i)) for i in range(n)]
    t0 = time.perf_counter()
    [t.start() for t in ts]; [t.join() for t in ts]
    dt = time.perf_counter() - t0
    wins = results.count('WIN')
    return wins, n - wins, dt, results

def main():
    print("=" * 64)

    # T1 — exclusive race on the private room
    wins, rej, dt, _ = race(50, ROOM, "[2026-09-01 14:00+00,2026-09-03 12:00+00)", True)
    ok1 = wins == 1
    print(f"T1 exclusive race (50 clients, 1 room):  wins={wins} rejects={rej}  "
          f"{dt*1000:.0f}ms   {'PASS' if ok1 else 'FAIL'}")

    # T2 — composite: 25 clients try DORM as a private room (exclusive),
    #      25 simultaneously try beds in it (non-exclusive), same dates.
    results = [None]*50
    ts = []
    per = "[2026-09-01 14:00+00,2026-09-03 12:00+00)"
    for i in range(50):
        excl = (i % 2 == 0)
        ts.append(threading.Thread(target=attempt, args=(DORM, per, excl, results, i)))
    [t.start() for t in ts]; [t.join() for t in ts]
    excl_wins = sum(1 for i,r in enumerate(results) if r=='WIN' and i%2==0)
    bed_wins  = sum(1 for i,r in enumerate(results) if r=='WIN' and i%2==1)
    # Legal outcomes: private sold (1,0) or beds sold (0,1..6) — never both.
    ok2 = (excl_wins == 1 and bed_wins == 0) or (excl_wins == 0 and 1 <= bed_wins <= 6)
    print(f"T2 composite race (private vs beds):     private={excl_wins} beds={bed_wins}"
          f"            {'PASS' if ok2 else 'FAIL'}  (never both)")

    # T3 — capacity: fresh dates, 40 clients race for 6 beds
    wins, rej, dt, _ = race(40, DORM, "[2026-10-01 14:00+00,2026-10-05 12:00+00)", False)
    ok3 = wins == 6
    print(f"T3 capacity race (40 clients, 6 beds):   wins={wins} rejects={rej}  "
          f"{dt*1000:.0f}ms   {'PASS' if ok3 else 'FAIL'}")

    # T4 — choke point: direct INSERT must be denied
    try:
        con = psycopg2.connect(DSN); cur = con.cursor()
        cur.execute("INSERT INTO space_occupancy (tenant_id,space_id,period,slot_ref,exclusive)"
                    " VALUES (%s,%s,%s::tstzrange,%s,true)",
                    (TEN, ROOM, "[2027-01-01,2027-01-02)", str(uuid.uuid4())))
        con.commit(); ok4 = False; msg = "INSERT SUCCEEDED (bad)"
    except psycopg2.Error as e:
        ok4 = e.pgcode == '42501'; msg = f"denied ({e.pgcode})"
    print(f"T4 choke point (direct INSERT as app):   {msg:>22}            "
          f"{'PASS' if ok4 else 'FAIL'}")

    # T5 — throughput: 500 non-overlapping bookings, 10 threads
    def worker(k, out):
        con = psycopg2.connect(DSN); cur = con.cursor(); n=0
        for j in range(50):
            d = k*50 + j
            cur.execute("SELECT record_occupancy(%s,%s,%s::tstzrange,%s,true)",
                (TEN, ROOM, f"[2028-01-01 {d//60:02d}:{d%60:02d}+00,2028-01-01 {d//60:02d}:{d%60:02d}:30+00)",
                 str(uuid.uuid4())))
            con.commit(); n+=1
        out[k]=n; con.close()
    out=[0]*10; t0=time.perf_counter()
    ths=[threading.Thread(target=worker,args=(k,out)) for k in range(10)]
    [t.start() for t in ths]; [t.join() for t in ths]
    dt=time.perf_counter()-t0; total=sum(out)
    print(f"T5 throughput:                           {total} bookings / {dt:.2f}s = "
          f"{total/dt:,.0f} commits/sec")
    print("=" * 64)
    allok = ok1 and ok2 and ok3 and ok4
    print("VERDICT:", "ALL PASS — the composite-slot constraint holds under concurrency."
          if allok else "FAILURES — investigate before ERD hardens.")
    return 0 if allok else 1

if __name__ == "__main__":
    raise SystemExit(main())
